### DISCUSSION TITLE
Welcome to eCAL Discussions!

### DISCUSSION THREAD

**Discussion by rex-schilasky** [eclipse-ecal/ecal#134] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by pratikmankawde** [comment]
Hi,
How can we guarantee quality of service at subscriber's end? Meaning, there should not be any loss of message.
Is there already a way to do it? Or is there a place to implement at receive queue?

**Comment by rex-schilasky** [comment]
Hi, eCAL is not providing any kind of quality of service because this would make the protocol more complicated and would finally decrease the overall performance. We run large software projects at Continental based on eCAL in the area of autonomous driving with many distributed software algorithms and hundreds of pub/sub interfaces.

If you face message drops on the subscribers side then there could be multiple reasons. One of them is that you subscriber is just to slow to process the incoming data send by the high frequency publisher. So even eCAL would use queuing it would not help in this use case, the queue would just overflow and you would finally work on outdate data from the queues end ..

IF you need a kind of queue because only SOMETIMES your subscriber is not able to process a message, then you can just write the incoming message in your own process internal queue. This can be done in a subscriber message receive callback.
In another thread you can then access that data, empty the queue or react on overflows. If this mechanism is still to slow to just copy the incoming message in a queue, then you should check to slow down your publication or put both components in ONE process and avoid ipc for that specific connection.

Can you give some more details on your use case ? We talk about C++ ? What is the send frequency of your publisher ? What size has your message ? Do you use a callback on publisher side ? Do you process the data INSIDE that callback or in another thread ?

**Reply by pratikmankawde** [reply]
Yes, it is a C++ project. I am creating following protocol buffer message.
```cpp
message change
{
  string variableName   = 1;     // Variable name, could be ~5-20 chars
  string oldValue       = 2;     // A float value converted to string
  string newValue       = 3;     // A float value converted to string
}

message Changes
{
    repeated change changes = 1;
}
```

I am sending 20 `Changes` objects one at a time .Each `Changes` object carry 200 `change` items.

At subscriber, I am just printing this received data to `std::out`. Currently, Subscriber prints the first message(carrying 1 `Changes` object, hence 200 `change` items) and then the last message. All intermediate messages are lost. This happened is all runs.
Alternatively when I tried to send each `change` data message one at a time(and not 200 together as single `Changes` message), subscriber only handled ~100 messages.
Publisher writes continuously, as soon as it has the data, without any wait/sleep. Adding a wait/sleep between publisher's write calls won't be acceptable. Let me know if you need any more information.

**Reply by rex-schilasky** [reply]
Okay, you are sending 20 messages of type ´´´Changes´´´ in one call (for loop ?) on the same topic and you expect that the subscriber is printing out all messages ? That will not work unfortunately. You can send 20 messages with 20 different topic names and they will all be received probably. If you send 20 messages (incredible fast :-)) on the same topic, the publisher will 20 times write this information in the same memory file and signal the subscriber (incredible fast again) for every single write action with a signal (semaphore). But here your subscriber is doing a slow print and has no chance to react on the updated information.

You can open 20 pub/sub connections (this will end in 20 shared memory files - no more overwriting here) or you repack you 20 ´´´Changes´´´ again in ONE message as a new protocol buffer type and send this new message at once.

You can also try to speedup your data processing and not print in the callback but only collect the incoming messages in your own queue and make the print asynchronously in another thread.

The logic is that one if you Send / Publish:

1. Publisher open / lock the memory file for that specific topic
2. Publisher copy your data into the file
3. Publisher unlocks the memory file
4. Publisher signal all connected subscriber that there is an updated information
5. Subscriber locks the file
6. Subscriber copy the content in it's process memory
7. Subscriber unlocks the memory file
7. Subscriber is calling your callback
8. Message is processed
9. Callback returns

If you have a slow operation inside the callback message processing (8), then you will not get all updates because you are blocking the callback operation and in the background the publisher is updating the memory file content.

**Reply by pratikmankawde** [reply]
Thanks for the quick reply!
This was only a test. In a real world scenario, the number of messages and the size of each message will be dynamic. Hence creating separate topics won't help much.
The only solution I see is if in step 6 of the logic you shared, the process memory becomes a queue. So Subscriber copies the message to a queue and then unlocks the memory. In step 8, it keeps running the callback in a separate thread. 

So, how much effort will it be to implement this? Can you please point me to the code sections where I can make this change? Unless if it's already in your to-do list for the project. Also, if you think this change might hits performance, this can be made optional. While creating instance of subscriber, a developer should be able to choose if he wants the queue based system or direct access system.
Let me know your thoughts on this.

**Reply by rex-schilasky** [reply]
Maybe there is misunderstanding. For sure you can try to change the ecal shared memory layer logic but I would not recommend this. Like I said, we had different kind of queuing and handshaking in older versions of eCAL and finally we removed all these mechanisms to make the core logic stupid simple and performant for all use cases (n:m pub/sub handling, dynamic changing payload sizes ..).

For your use case the easiest solution is. Just add a queue logic (thread safe) to your application message callback and copy the incoming messages into it. So you should not loose that many messages. In a separate thread you can access that container and process (print) your content.

If you want to have a minimal handshake in the memory file handling you can modify the parameter ```memfile_ack_timeout``` in the ecal.ini file. This parameter is the maximum time in milliseconds for the publisher waiting for a ack signal from the subscriber that the content is processed. But I do strongly recommend to better change your application logic like I mentioned above.

**Reply by pratikmankawde** [reply]
I understand that there can be potential risks of adding the queue mechanism in the eCal code than on my application side. But, on the application side, there is no guarantee on how much time the data copy will take in the callback function of Subscriber. Given that the lock on shared memory will already be released, Publisher will overwrite it. Only if we do this copy to queue before we release the lock on shared memory, can we guarantee that there won't be any message lose. As per our application requirements, we can't accept any message drop.
If I got your idea about `memfile_ack_timeout` correctly, I think increasing it won't help in this case, because Subscriber will read the message from shared memory and then release the lock instantaneously. Then the extra operation of copying this message to my application side queue will take some more time. Since the shared memory is unlocked, publisher will keep publishing to it. The lock should persist until the message is added to the queue, irrespective of where I implement the queuing mechanism.

**Reply by rex-schilasky** [reply]
Yes, the ini file parameter will just prevent the publisher to overwrite the memory file content BEFORE the subscriber has copied out the content in it's own process memory.

If you really can not accept any message lost then maybe eCAL is the wrong ipc framework for you. Our design decision was maximum performance over hundred percent reliability, so let's call it "best effort" :-).

You can try out the eCAL client/server pattern, that is based on tcp/ip and will guarantee reliability so far. Or you create your own sync mechanism like a feedback message, but that makes not that much sense. You can btw. react on dropped messages by adding an event callback like shown here
https://github.com/continental/ecal/blob/master/samples/cpp/person/person_rec_events/src/person_rec_events.cpp

Another option is to use the eCAL iceoryx build (we are currently porting to iceoryx 1.0.0, there is an open PR). Iceoryx has some built in queueing you can read about it here
https://iceoryx.io/v1.0.0/

And finally for sure you can fork eCAL and completely implement your own memory transport layer or modify the existing one. Just check the implementation details here
https://github.com/continental/ecal/tree/master/ecal/core/src/io

**Reply by pratikmankawde** [reply]
I will look through these. Thanks for sharing. :-)
Do you know of any other framework what supports pub/sub with QOS? I loved the the 'topic' feature in eCal. It's almost like zero setup and all nodes in the network can talk to each other so easily. We want exactly that.

**Reply by rex-schilasky** [reply]
Did you checkout ROS2 - https://docs.ros.org/en/rolling/ ? Or DDS systems like CycloneDDS - https://github.com/eclipse-cyclonedds/cyclonedds ? They support all kind of QOS and Pub/Sub for sure !

**Reply by pratikmankawde** [reply]
Thanks for sharing. I will check them.

**Reply by raphaelzulliger** [reply]
I'm also interested in eCAL. I found this an interesting discussion, sicne I kind of have the same questions.

Two things:

1. Why not "create our own sync mechanism" by having two topics forth and back
2. Why don't services help with QOS?

## In more detail

1. What I do not understand is why you (@rex-schilasky) mention "Or you create your own sync mechanism like a feedback message" but you add that "but that makes not that much sense":

- Having two topics, e.g. "Request" and "Acknowledge"
- The sender only sends the next messages once it received an acknowledge via "Acknowledge" topic

By this, it would be implementable, wouldn't it? Are you saying that this "doesn't make sense" because then we have to implement a timeout/retry/flush/... logic ourselves (which is not as trivial as it first seems). And you consider that bad, since there are other frameworks out there already implementing this?

What I like about eCAL is the distributed/multicast feature along with its protocol agnostic feature: Protobuf vs. whatever. So maybe the "drawback" of implementing our own QOS on top of it would be a price we'd like to pay?!

2. Initially, having read about "Services", I thought I'll get the wanted QOS, kind of. But you seem to be skeptical about that one as a viable solution as well. Why? (The docs on Services seems a bit sparse so I may not have understood how they actually work)

**Reply by rex-schilasky** [reply]
Hi Raphael,

> 1. What I do not understand is why you (@rex-schilasky) mention "Or you create your own sync mechanism like a feedback message" but you add that "but that makes not that much sense":
> 
> * Having two topics, e.g. "Request" and "Acknowledge"
> * The sender only sends the next messages once it received an acknowledge via "Acknowledge" topic
> 
> By this, it would be implementable, wouldn't it? Are you saying that this "doesn't make sense" because then we have to implement a timeout/retry/flush/... logic ourselves (which is not as trivial as it first seems). And you consider that bad, since there are other frameworks out there already implementing this?
> 

Yes, to implement this will increase complexity and decrease performance for sure. Think about connections of multiple publisher to multiple subscriber running distributed on different hosts (and so using different transport mechanisms). Maybe some of them should process every single message but some of them are slow processing for some reasons and not interested on every message (visualization nodes ..).

So, eCAL itself is not providing a feedback mechanism to guarantee the highest possible "best effort" performance. But for sure such a mechanism can be implemented on user level, you are right - but then it's your job to not slow down you whole system ;-).

The two eCAL C++ samples ```latency_snd``` and ```latency_rec_cb``` are using such a feedback connection by the way to fully snyc publisher and subscriber without loosing a single message.

> What I like about eCAL is the distributed/multicast feature along with its protocol agnostic feature: Protobuf vs. whatever. So maybe the "drawback" of implementing our own QOS on top of it would be a price we'd like to pay?!

Thank you for your feedback, we really appreciate this ! To use different protocols was (besides performance) another design goal.

> 1. Initially, having read about "Services", I thought I'll get the wanted QOS, kind of. But you seem to be skeptical about that one as a viable solution as well. Why? (The docs on Services seems a bit sparse so I may not have understood how they actually work)

Services are a different communication pattern and no replacement for a publish subscribe pattern. For large systems with hundreds or thousands of data connections and many of them between one sources to one or multiple sinks, services are no alternative IMHO. They create latency everywhere and can cause bad timeouts or deadlocks in some scenarios.

**Comment by AntoineBassoul** [comment]
Hello eCAL !

I know ecal is great to use as a real time pub/sub library.
I'd be interested in directly writting & reading eCAL data from ecal rec files to do deterministic simulation and build custom applications that read ecal data as fast as possible (I hope that's somehow clear, i'd like to do something similar to what is done in the "yaaf simulator binding").
I didn't see any documentation regarding this, is there something available in the API ?

Thanks and Best,
Antoine (now at https://navya.tech/en/)

**Comment by rex-schilasky** [comment]
Hi and thank you for the nice words !

The eCAL HDF5 C++ and python API is open source (YAAF is an internal frame generator).

Here is a sample how to read a measurement (message encoding google protobuf) in C++:
https://github.com/continental/ecal/blob/master/samples/cpp/measurement/measurement_read/src/measurement_read.cpp

and this one for Python:
https://github.com/continental/ecal/blob/master/samples/python/measurement_read/measurement_read.py

This is the google unit test:
https://github.com/continental/ecal/blob/master/testing/contrib/ecalhdf5/hdf5_test/src/hdf5_test.cpp

So finally you can integrate the C++ HDF5 measurement interface into your algorithm frame and stimulate your interface in a deterministic way independently from real-time.

Best, ReX

**Comment by 66hades** [comment]
Hello, 
I'm a Systems Test Engineer learning to use eCAL on development vehicles to evaluate ADAS features.
I tried to start a new discussion thread, but apparently do not have permissions to do that, so I'm piggybacking on this thread.
Question: Is there an eCAL equivalent to ROS's rqt_graph? Oftentimes it is more understandable to see a node graph of various processes and the topics they're using. I can only find the Monitor App which only provides lists, and unless I'm missing something not a graph.
Thanks, Brian

**Reply by rex-schilasky** [reply]
Hi Brian,
currently there is no equivalent tool like rqt_graph for eCAL. Since the monitoring information is quite simple to snapshot via python script one option is to use python packages for node plotting and create a graph out of this. Would that be a help for you ? Maybe we can add a script showing how to do so ...
Best, ReX

**Reply by 66hades** [reply]
Hello Rex,
That would be helpful! I'm familiar with C++ & Python, any hints on where to get the current list of nodes & topics, a good graph visualizer repo, and a script to incorporate it into eCAL would be extremely appreciated.
Thanks, Brian

**Reply by rex-schilasky** [reply]
Hi Brian, I made a first prototype for generating a graph for the ecal monitoring. Just take a look here [ecal_mongraph](https://github.com/ecal-io/ecal_mongraph). Would be nice if we could improve that together.

**Comment by davidtwomey** [comment]
Hi,

Relatively new to ecal but very much enjoying the development experience so far!
I have a quick question:  
Before creating a service, I want to  check if a service with the same service name already exists (on the same host).
Is there a recommended way to do this?

Thanks

**Reply by rex-schilasky** [reply]
Hi,

you can use the monitoring API shown in this sample [monitoring_reg](https://github.com/continental/ecal/blob/master/samples/cpp/monitoring/monitoring_reg/src/monitoring_reg.cpp). In the next version 5.10 you can just start a client and attach a "connect event" to check if there is a service with the same name connecting to you. Like shown [here ](https://github.com/continental/ecal/blob/feature/service_events/samples/cpp/services/math_client/src/math_client.cpp).

Enjoy eCAL, ReX

**Comment by long131ss** [comment]
hi, I integrated the ecal framework into Kinetic ros,and it can works on my system.
But I found the cpu usage ratio is high. for example, I just subsciber a topic /scam with the frequency 10hz, the message size is 3kb,
another topic /map ,with frequency 0.2hz,message size is 35kb, I can find the cpu ratio is about 10%, as a comparison, I also tested another pub/sub framework zero-zmq ,I found the cpu usage is very low ,about 3%。
Question: do you test the cpu usage? or although ecal with share memory is fast ,but it consumes the cpu more much?
If I want run the ecal framework with relatively low cpu usage ,what can i do?

**Comment by rex-schilasky** [comment]
> hi, I integrated the ecal framework into Kinetic ros,and it can works on my system.
> 
> But I found the cpu usage ratio is high. for example, I just subsciber a topic /scam with the frequency 10hz, the message size is 3kb,
> 
> another topic /map ,with frequency 0.2hz,message size is 35kb, I can find the cpu ratio is about 10%, as a comparison, I also tested another pub/sub framework zero-zmq ,I found the cpu usage is very low ,about 3%。
> 
> Question: do you test the cpu usage? or although ecal with share memory is fast ,but it consumes the cpu more much?
> 
> If I want run the ecal framework with relatively low cpu usage ,what can i do?
> 
> 

Can you please share some details about your platform ?

**Reply by long131ss** [reply]
Hi, thanks for your reply. I tested it  on the platform rockchip.x3399 with arm x64 and 1G memory. do you have any suggestions for reduce of the cpu usage?

**Reply by rex-schilasky** [reply]
Can you please make a test without publication. Just create all entities (pub/sub) and start your processes ?
There is a background registration layer and I want to figure out how much it consumes on your platform.
We run the middleware mainly on large server systems with some more cpu power ...

**Reply by long131ss** [reply]
maybe the ecal framework is more suite to the large message transport ,such as automatic driving . And now i want to apply it to my Robot, the sensor message size is less than 256kb.

**Reply by long131ss** [reply]
> Can you please make a test without publication. Just create all entities (pub/sub) and start your processes ?
> There is a background registration layer and I want to figure out how much it consumes on your platform.
> We run the middleware mainly on large server systems with some more cpu power ...

thanks for your kind reply. i will tell you test result once i get back my work. Because our system is strictly to the cpu usage, so it is important to me.

**Reply by long131ss** [reply]
Hi, done as you said, just create all entities and no publication, the CPU usage is about 5%.

**Comment by davidtwomey** [comment]
Hi @rex-schilasky,

Another quick question for you. Am I right in saying there is currently no way for a python Client to interact with a c++ protobuf Service?
The only way would be to deserialize the protobuf manually and use string methods?

Thanks

**Reply by rex-schilasky** [reply]
Hi @davidtwomey,

yes that's right currently. The protobuf service layer is not implemented on python API for now.

Best, ReX

**Reply by davidtwomey** [reply]
Thank you for the prompt responses!

With a protobuf service client (for example the "ping" sample), is there a way to target a `service_name` explicitly?

For example you may have two "Pong" servers but with different service_names, how would i make a call on ServerA but not ServerB?
Or is that not supported for Proto services?

For example:

```cpp

// create ping service client
eCAL::protobuf::CServiceClient<PingService> ping_client;

ping_client.Create("ServerA"); // Connect to "ServerA" ONLY by service name (?)

// waiting for connection omitted...

// Call just PingServer A
Ping ping;
ping_client.Call("Ping", ping);
```

Thanks,

David

**Reply by rex-schilasky** [reply]
Hi David,

that is a very good question. This use case is not covered by the current protobuf service implementation. This looks like this

Server Side:

```cpp
      /**
       * @brief Create service.
       *
       * @param service_  Google protobuf service instance.
       *
       * @return  True if successful.
      **/
      bool Create(std::shared_ptr<T> service_)
      {
        if (service_ == nullptr) return false;
        m_service = service_;

        const google::protobuf::ServiceDescriptor* service_descriptor = service_->GetDescriptor();
        Create(service_descriptor->full_name());       << the protobuf service name is used to name the server
```

Client Side:

```cpp
    /**
     * @brief Google Protobuf Client wrapper class.
    **/
    template <typename T>
    class CServiceClient : public eCAL::CServiceClient
    {
    public:
      /**
       * @brief Constructor.
       *
      **/
      CServiceClient()
      {
        Create(T::descriptor()->full_name());      << the protobuf service name is used to name the client
      }
```

I will open an issue and we will find a solution here.

Best, ReX

**Reply by davidtwomey** [reply]
Right I see! Yes perfect, let’s start from an issue and see how we can resolve

**Reply by rex-schilasky** [reply]
Set an issue now (#493). Let's discuss this there ...

**Reply by rex-schilasky** [reply]
Going to master currently ... See "ping" service samples how to use it. 

24 hours service ;-)

Good luck, ReX

**Reply by davidtwomey** [reply]
Amazing - thanks Rex!

**Comment by davidtwomey** [comment]
Another question for you @rex-schilasky :)

I have created a very simple pub/sub scenario where I am sending proto messages (via shm on the same host).
However, if the publisher sends 2 (or more) messages sequentially with no sleep, the subscriber drops at least one of these messages.
This is even the case when the subscriber callback is empty!

(Ignoring for a sec the point that if you are sending n messages sequentially without sleep, why not batch them)
Are there any settings I can configure to ensure I do not (or at least minimise) dropped messages? Possibly:
- Change sub/pub buffer size?
- Change QoS?
- Switch to zero-copy?
- Have to revert to tcp?

Or is the ONLY solution to add a small sleep between publishing? 1ms works for example

Many thanks!

David

**Reply by rex-schilasky** [reply]
Hi David,

there is a global configuration setting (ecal.ini) where you can define a handshake with timeout between pub/sub connections for local communication. See #441.

Maybe we need to better document this option ;-).

Best, ReX

**Reply by davidtwomey** [reply]
Great, I tried with setting it to 100ms, that seems to resolve the issue!

What exactly does enforcing a handshake timeout do here? 
Is lower always better? Can I, for example set this down to 1? Or are there other considerations here?

**Reply by rex-schilasky** [reply]
It couples the publisher with a defined maximum timeout to the connected subscribers memory reader.

Not to the callback processing time, only to time the subscriber needs to make a copy of the memory files content.

So higher values will not really decrease performance.

There are special cases, for example if you activate zero copy additional THEN the subscriber will work on the memory file content directly and so blocking the publisher until the callback function returns.

**Reply by davidtwomey** [reply]
Understood thanks! 
Is there a specific reason the default value is set to 0?

**Reply by Kerstin-Keller** [reply]
The thing is, this setting can lead to unintended side effect, which are mainly relevant in 1:N scenarios (1 publisher, N subscribers).
If any of the N  subscribers is blocked from receiving their message (which until the next release happens if arbitrary time is spend in a callback, even of another topic subscription), the sender is blocked from sending at a given frequency. 

In big scenarios, a few framedrops for any subscribers is much more acceptable than the whole system experiencing delays.

**Comment by harlenli2022** [comment]
Hi @rex-schilasky,

Nice to contact you, we are planning to use eCAL in our Cockpit and ADAS system, and currently we had ported eCAL to arm linux successfully.
Since more and more systems are to use QNX hypervisor solution, so we have to support eCAL in QNX system also.
Currently I'm trying to build the QNX version, but seems some third party dependencies (Asio, etc) don't support QNX very well.

Has eCAL ported successfully before to QNX system? it would be great helpful for our estimation, hope we are not the first one to do this...

Thanks

**Reply by rex-schilasky** [reply]
Hi,

I'm very curious for what company you are planning to use eCAL, would be interesting for my developer team :-).

Regarding your question. Yes, together with MATHWORKS we ported and run eCAL on top of QNX. Please take a closer look to the [ecal-toolbox](https://github.com/mathworks/ecal-toolbox) repo for simulink first.

Best, ReX

**Comment by harlenli2022** [comment]
Hi,

I met issues on network configuration when using eCAL to communicate between two PCs in different sub-net (e.g. 192.168.1.x and 192.168.2.x).
Following the wiki page, I had added route 239.0.0.0 already, after then two PCs in same sub-net (192.168.1.xx) can exchange info via eCAL successfully, but in case of different sub-net, server/client (same to pub and sub) can't identify each other...

Could you support to confirm if special configuration needed for my issue? 

Thanks in advance.

**Reply by harlenli2022** [reply]
@rex-schilasky,
what's more, I also built eCAL in MacOS, everything seems ok except ecal rec application can't be launched with below protobuf errors, have your team occurred same issue? thanks.
> [libprotobuf ERROR /Users/HarlenLI/workspace/ecal/thirdparty/protobuf/src/google/protobuf/descriptor_database.cc:120] File already exists in database: ecal/pb/host.proto
> [libprotobuf FATAL /Users/HarlenLI/workspace/ecal/thirdparty/protobuf/src/google/protobuf/descriptor.cc:1356] CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size):
> libc++abi: terminating with uncaught exception of type google::protobuf::FatalException: CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size):

**Reply by Kerstin-Keller** [reply]
@FlorianReimold maybe you can answer on UDP settings?

As for protobuf, which commit did you build, did you build one of the tags or the current master? 
Usually you get these type of protobuf errors when any proto datatype is compiled multiple times into the program. I am not aware that we are doing any such thing in eCAL Rec. If it occurs in Macos, I would think it occurs on Ubuntu as well. What protobuf version are you using in the build?

**Reply by FlorianReimold** [reply]
Hi Harlenli,

Can you create either create a new discussion or an issue with your question? In the past we had misconfigured github, so everybody created their questions as a comment to this discussion as they were not able to create discussions (So it is basically our fault that this discussion looks as it looks 😑). But now creating discussions should be possible for all users and a proper new GH Discussion will better help other users in the future.

**Reply by harlenli2022** [reply]
@FlorianReimold 
Sorry for my mistake, I had created separate discussion instead, let's switch to the new discuss then.
